### ISSUE TITLE
mu index runs really slow on openbsd

### ISSUE THREAD

**Issue by mdom** [djcb/mu#1335]
## Expected or desired behaviour

Running mu index shouldn't take minutes for a simple update.

## Actual behaviour

I ran my first mu index over night so i did not notice how long it tool. The next day I wondered why every mu index would take several minutes to run. To be sure that I didn't had a corrupt index, I rebuild the index and noticed that it took forever. It seems to average on 35 msg per second. I have about 120.000 mails. Just writing those files to /dev/null takes about 1.5 minutes.

## Steps to reproduce

Just run mu index --rebuild

## Versions of mu, mu4e/emacs, operating system etc.

mu 1.0 on openbsd 6.4

## Any other detail

I already mounted my disk with softdep and noatime.

**Comment by djcb** [comment]
That's indeed quite slow. Did it ever go faster on your system?

You can speed up subsequent index operations with `--lazy-check` and `--no-cleanup`, and perhaps check `.noupdate` (all explained in the `mu-index` manpage).

If that doesn't help, some profiling information on where the time is spent would help.

**Comment by mdom** [comment]
This is the first time I try mu, but I used mairix before and all my mails were indexed in under two minutes. Not that I expect the same performance from both programs, just as comparison to get a feeling how fast my system can read those mails. I checked ktrace but didn't notice anything obvious. I check with gprof when I have same spare time.

**Comment by djcb** [comment]
Yeah, please share any profiles you have. I suspect it might be something OpenBSD-specific.

**Comment by mdom** [comment]
I compiled mu with -g -pg but somehow i'm not able to get a gmon.out. That's very embarrassing! I have a 14 MB cleaned up kdump file, do you want that? Not sure how helpful that would be? Sorry.

**Comment by djcb** [comment]
That wouldn't help much, I'm afraid, unless you can attach some textual output -- something like what Linux' `perf record` could deliver (it's been a while since I developed on OpenBSD, so I have become a bit Linux-centric!)

**Comment by mdom** [comment]
I'm not sure there's anything like `perf record` for openbsd (but if there is, i'm happy to run it and provide the output). So i did the next best thing and compiled mu for gprof. This is the output of my short run, i'll attach the gprof output. Hope that helps!

```
balin$ ./mu/mu index -m /home/mdom/Maildir/
indexing messages under /home/mdom/Maildir/ [/home/mdom/.mu/xapian]                                                              
| processing mail; processed: 975; updated/new: 975, cleaned-up: 0

mu: shutting down gracefully, press again to kill immediately
- processing mail; processed: 1050; updated/new: 1050, cleaned-up: 0
elapsed: 134 second(s), ~ 7 msg/s
```

**Comment by mdom** [comment]
I'll also add the output of `kdump -R -m0`. It's not a complete run, but i hope it helps!
[kdump.txt](https://github.com/djcb/mu/files/2930029/kdump.txt)

**Comment by manphiz** [comment]
I'm experiencing the same problem on OpenBSD 6.6-stable.  My guess is that this may have something to do with the file system.  On average FFS is slower than EXT4.  As the message count goes up, the xapian index file grows really large and any updates becomes really slow.  An untested wild guess: if the xapian index can be segmented to small chunks it may help.  But don't know whether that's supported.

**Comment by handstad** [comment]
This is probably not relevant for you, but it might help others: I had big problems with speed on MacOS, indexing took around 2 minutes. Flags `--lazy-check` or `--nocleanup` did not help. I found out that I had a lot of mails (22 thousand) marked for deletion (mu4e shortcut `d`) and that these were somehow getting synced every time I updated with mbsync and were not expunged locally or on my remote server. Mu then "cleaned them up" every time (see attachment). 
![mu_index_timing_3](https://user-images.githubusercontent.com/1928673/72991018-70f8fd00-3df1-11ea-875c-f901b9cfceb5.png)

I solved this by deleting these emails in my local Trash manually with rm in the shell, then I ran `mbsync -d --expunge-master --expunge-slave myaccountname` and now both mbsync and mu index runs in around a second or less. I also delete messsages directly (mu4e shortcut `D`), because if I use `d` messages are just moved to Trash and stay there. I attach my .mbsyncrc file also.
[mbsyncrcfile.txt](https://github.com/djcb/mu/files/4103461/mbsyncrcfile.txt)
